File: 2015/day6.jl
function parsecoordinates(instruction::String)
    numbers = parse.(Int, match(r"(\d+),(\d+) through (\d+),(\d+)", instruction).captures)
    x1, y1, x2, y2 = numbers .+ 1  # broadcast +1 across elements to account for 0 indexing
    return x1:x2, y1:y2
end

function lighttoggle!(instruction::String, lights::BitMatrix)
    rows, cols = parsecoordinates(instruction)
    lights[rows, cols] .= .!lights[rows, cols]  # Broadcasts the not operator (!) across the selected rows and cols
end

function lightoff!(instruction::String, lights::BitMatrix)
    rows, cols = parsecoordinates(instruction)
    lights[rows, cols] .= false
end

function lighton!(instruction::String, lights::BitMatrix)
    rows, cols = parsecoordinates(instruction)
    lights[rows, cols] .= true
end

# part 2

function lightbrighten!(instruction::String, brightlights::Matrix)
    rows, cols = parsecoordinates(instruction)
    brightlights[rows, cols] .+= 2
end

function lightdim!(instruction::String, brightlights::Matrix)
    rows, cols = parsecoordinates(instruction)
    brightlights[rows, cols] .= max.(brightlights[rows, cols] .- 1, 0)
end

function lightincrease!(instruction::String, brightlights::Matrix)
    rows, cols = parsecoordinates(instruction)
    brightlights[rows, cols] .+= 1
end


function main()
    lights = falses(1000, 1000)
    brightlights = zeros(Int, 1000, 1000)

    for line in readlines("2015/inputs/day6.txt")
        if startswith(line, "toggle")
            lighttoggle!(line, lights)
            lightbrighten!(line, brightlights)
        elseif startswith(line, "turn on")
            lighton!(line, lights)
            lightincrease!(line, brightlights)
        elseif startswith(line, "turn off")
            lightoff!(line, lights)
            lightdim!(line, brightlights)
        end
    end

    litlights = count(lights)
    println("There are $(litlights) lights lit")

    totalbrightness = sum(brightlights)
    println("The total brightness is $(totalbrightness)")
end

main()

File: 2015/day6.m
function day6
    lights = false(1000, 1000);
    brightLights = zeros(1000, 1000);

    instructions = readlines("inputs/day6.txt");
    for i = 1:length(instructions)
        line = strtrim(instructions(i));

        if startsWith(line, "toggle")
            lights = lightToggle(line, lights);
            brightLights = lightBrighten(line, brightLights);
        elseif startsWith(line, "turn on")
            lights = lightOn(line, lights);
            brightLights = lightIncrease(line, brightLights);
        elseif startsWith(line, "turn off")
            lights = lightOff(line, lights);
            brightLights = lightDim(line, brightLights);
        end
    end

    litLights = sum(lights, "all");
    fprintf("There are %d lights lit\n", litLights);

    totalBrightness = sum(brightLights, "all");
    fprintf("The total brightness is %d\n", totalBrightness);
end


function [rows, cols] = parseCoordinates(instruction)
    tokens = regexp(instruction, '(\d+),(\d+) through (\d+),(\d+)', 'tokens');
    numbers = cellfun(@str2double, tokens{1});
    x1 = numbers(1) + 1;
    y1 = numbers(2) + 1;
    x2 = numbers(3) + 1;
    y2 = numbers(4) + 1;

    rows = x1:x2;
    cols = y1:y2;
end

function lights = lightToggle(instruction, lights)
    [rows, cols] = parseCoordinates(instruction);
    lights(rows, cols) = ~lights(rows, cols);
end

function lights = lightOff(instruction, lights)
    [rows, cols] = parseCoordinates(instruction);
    lights(rows, cols) = false;
end

function lights = lightOn(instruction, lights)
    [rows, cols] = parseCoordinates(instruction);
    lights(rows, cols) = true;
end

function brightLights = lightBrighten(instruction, brightLights)
    [rows, cols] = parseCoordinates(instruction);
    brightLights(rows, cols) = brightLights(rows, cols) + 2;
end

function brightLights = lightDim(instruction, brightLights)
    [rows, cols] = parseCoordinates(instruction);
    brightLights(rows, cols) = max(brightLights(rows, cols) - 1, 0);
end

function brightLights = lightIncrease(instruction, brightLights)
    [rows, cols] = parseCoordinates(instruction);
    brightLights(rows, cols) = brightLights(rows, cols) + 1;
end